File: Python/14get_node_value.py


# iterative
# n = number of nodes
# Time: O(n)
# Space: O(1)

def get_node_value(head, index):
  count = 0
  current = head
  while current is not None:
    if count == index:
      return current.val

    current = current.next
    count += 1

  return None

# ---------------------------------------------
# iterative.  stopping at index
# n = number of nodes
# Time: O(n)
# Space: O(1)

def get_node_value(head, index):
  pass # todo
  current = head
  i = 0

  while current is not None and i < index:
    current = current.next
    i += 1

  return current.val if current else None



# -----------------------------------------

# recursive
# n = number of nodes
# Time: O(n)
# Space: O(n)


def get_node_value(head, index):
  if head is None:
    return None

  if index == 0:
    return head.val

  return get_node_value(head.next, index - 1)

File: Javascript/14getNodeValue.js


// RECURSIVE
// n = number of nodes
// time:  O(n)
// space: O(n)
const getNodeValue = (head, index) => {
  // todo
  if (!head) return null;
  if (!index) return head.val;
  return getNodeValue(head.next, index-1);
};


// ------------------------------------- //
// ITERATIVE
// n = number of nodes
// time:  O(n)
// space: O(1)
const getNodeValue = (head, index) => {
  // todo
  let current = head;
  
  while (current) {
    if (index === 0) return current.val;
    index--;
    current = current.next;
  }
  
  return null;
};